### DISCUSSION TITLE
Asp net core port mapping

### DISCUSSION THREAD

**Discussion by Simkiw** [dotnet/dotnet-docker#4393] [Q&A] [answered]
Hi,  

I have been looking for hours why my web api does not listen to my container port other than 80.  
It turns out there is an environment variable already set with its value **ASPNETCORE_URLS=http://+:80**, and no matter what I use in my docker run command, it would still be listening to that port.  

I found this information on [stackoverflow](https://stackoverflow.com/questions/55215279/how-to-expose-an-asp-net-core-docker-container-to-a-public-url) however, when i tried to find it by myself, I couldn't.  

Usually, the docker image decription in the docker hub presents the necessary information that would help run the container, in my case, it wasn't : **docker run -p hostPort:containerPort** works only if I map it to 80, otherwise there is an extra step to go over.  

As a developer who would like to write .net apps and containerize them, what are the things I should configure and how can i find the information, like that environment variable which in my opinion is very important.  

Anyway, I just try to understand why is it necessary to go through that env variable and how I can find the information again.  
Thank you

**Comment by mthalman** [comment, marked as answer]
You should be able to explicitly set the port via the `docker run` command by setting the `ASPNETCORE_URLS` variable. Here I set the app in the container to listen on port 9999 and map that to port 8000 on my host machine:

```
docker run --rm -it -p 8000:9999 -e ASPNETCORE_URLS=http://+:9999 mcr.microsoft.com/dotnet/samples:aspnetapp
```

**Reply by Simkiw** [reply]
Yes, thank you for the confirmation.
But this command  
`docker run -p 8000:9999 mcr.microsoft.com/dotnet/samples:aspnetapp`   
In this case, the container port 9999 is ignored and will be listening on 80 because by default, that **ASPNETCORE_URLS** is already set to 80.  

- I just wonder why we must set both, the container port and the environment variable, since they both have the same value? (why setting the container port wouldn't be enough?)
- And how can i find the information regarding that **ASPNETCORE_URLS**? (I have checked out the microsoft docs, docker hub, I haven't seen any)  

Maybe it's by design, maybe that's how .net is configured, maybe the docs is lacking, i don't know, i'm just trying to understand, maybe I missed the information because of my lack of expertise too.

**Reply by mthalman** [reply]
> * I just wonder why we must set both, the container port and the environment variable, since they both have the same value? (why setting the container port wouldn't be enough?)

Because Docker and ASP.NET Core are independent of each other. ASP.NET Core can run outside of a Docker container and thus needs to have a way to configure its endpoint that isn't tied to Docker. Conversely, Docker knows nothing about ASP.NET Core and has its own way of specifying which port of the container to expose to the host.

> * And how can i find the information regarding that **ASPNETCORE_URLS**? (I have checked out the microsoft docs, docker hub, I haven't seen any)

It doesn't necessarily have to be the `ASPNETCORE_URLS` variable. That's just a convenient way to set it. There are [other ways to instruct ASP.NET Core which endpoint to use](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints).

**Reply by Simkiw** [reply]
Alright.  
Thank you again.  

It's important because ... like I said, I naively containerized the app. And, between the `launchSettings.json` and the `containerPort` in the docker run command, I couldn't understand how my container was mapped to 80 and why I could not override it.  

I can say that I feel comfortable with .Net but there are still things i'm not aware of.  
Just to say I appreciate the provided details.

**Reply by Simkiw** [reply]
> It doesn't necessarily have to be the ASPNETCORE_URLS variable. That's just a convenient way to set it. There are [other ways to instruct ASP.NET Core which endpoint to use](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints).

But that would still make me use the .net image environment variable to set the container listening port, right?  
(I make the distinction between the asp net core app and the container of that asp app)

**Reply by mthalman** [reply]
You would still need to set the Docker port argument (`-p`) regardless of how you end up configuring the endpoint of the ASP.NET Core app.

**Reply by Simkiw** [reply]
Of course.
